### PR TITLE
Include Rinq's traceID in opentracing span tags.

### LINKED ISSUE #128: Include Rinq's "trace ID" in opentracing span tags.
I initially excluded this because the traceID is typically equivalent to the first message ID in a tree of requests - which is visible in an OpenTracing graph now. However, it's possible for an application to set a custom trace ID (via the `trace` package), and this information is not displayed anywhere.

Such a custom trace ID could still be very useful if one half of a peer-peer interaction does not have an OpenTracing implementation, for example.

### PR DESCRIPTION
**Pull request by koden-km**
Fixes #128

### CODE REVIEW

File: src/internal/opentr/common.go [hunk 1]
```diff
@@ -1,8 +1,18 @@
...
+	"github.com/opentracing/opentracing-go/log"
+)
 
 var (
 	successEvent = log.String("event", "success")
 	errorEvent   = log.String("event", "error")
 )
+
+// AddTraceID configures span s to have traceID set to the given id.
+func AddTraceID(s opentracing.Span, id string) {
```
Comment (jmalloc) [inline src/internal/opentr/common.go:14]: Please add tests for this function.

**Review by jmalloc — DISMISSED** [review]
There are several other places where this change needs to be made.

Firstly, look for other uses of `command.Invoker` and `command.Notifier`.  There's probably no others for the notifier, but the command invoker and server are both used in the `remotesession` package for the `_sess` namespace, and also use perform tracing ops.

Finally, there's also some tracing performed on the "destination" end for both notifications and commands, which if I remember correctly is performed inside a closure in `localsession.Session.Listen` and `localsession.Session.SetAsyncHandler()`, respectively.

**Comment by koden-km** [comment]
I'm not sure how to get a `traceID` in local session SetAsyncHandler?
`src/internal/localsession.session.go`
`func (s *Session) SetAsyncHandler(h rinq.AsyncHandler) error {`

**Comment by koden-km** [comment]
I'm not sure about adding the trace to the span for notifier here:
`src/rinqamqp/internal/notifyamqp/notifier.go`
`func (n *notifier) NotifyUnicast(` and `func (n *notifier) NotifyMulticast(`.

I don't have the span it gets packed?

**Comment by jmalloc** [comment]
> I'm not sure how to get a traceID in local session SetAsyncHandler?

The closure inside is passed a `ctx`, which is guaranteed to contain the trace ID by that point, so you can fetch it with `trace.Get(ctx)`.

**Comment by jmalloc** [comment]
> I'm not sure about adding the trace to the span for notifier here:
> src/rinqamqp/internal/notifyamqp/notifier.go
> func (n *notifier) NotifyUnicast( and func (n *notifier) NotifyMulticast(.

I don't think you'll need to because the methods of `localsession.Session` that use these will have already done it.

**Comment by koden-km** [comment]
OK I have added the trace id to the `SetAsyncHandler` using `trace.Get(ctx)`.
I checked the NotifyXxx methods and you are correct, they already have the trace set earlier in the Session.
I think that is all of those covered now.

I wasn't sure whether to add the extra blank line of spacing for those calls to `opentr.AddTraceID()`. I just tried to match the existing style and there seemed to be a blank line before the error check lines. I can remove that extra blank line if you prefer them packed tighter?

**Comment by jmalloc** [comment]
> I can remove that extra blank line if you prefer them packed tighter?

I am trying to care less about this kind of trivial stuff.  If you run `make prepare` and it doesn't change it, then it's good!

**Review by jmalloc — COMMENTED** [review]
Trying review comment without approving or requesting changes.

**Review by jmalloc — DISMISSED** [review]
Requesting changes.

**Review by jmalloc — CHANGES_REQUESTED** [review]
There are a few more places that need to have the trace ID added to the span:

- `rinqamqp.peer.Listen()` has a closure similar to `Session.SetAsyncHandler()`
- `remotesession.server` also creates spans

I *think* if you look for everywhere that `opentr.Setup<something>()` is called that will identify all the locations.

**Comment by koden-km** [comment]
OK I will fix those up now.

**Comment by jmalloc** [comment]
Before I merge this, I'd like for you to spin up Jaeger under Docker and try out each of these scenarios.

http://jaeger.readthedocs.io/en/latest/getting_started/#all-in-one-docker-image

There's no Rinq environment variable for configuring OpenTracing, so you'll need to configure the client yourself.

This is the client we use for Jaeger: https://github.com/jaegertracing/jaeger-client-go

That Jaeger client is just one possible implementation of the opentracing interfaces that Rinq uses.

**Pull request merged**